pkg/apiserver/rest/dualwriter_mode1.go

```diff
@@ -27,36 +34,197 @@ func NewDualWriterMode1(legacy LegacyStorage, storage Storage) *DualWriterMode1
-	ctx = klog.NewContext(ctx, d.Log)
-	return d.Legacy.DeleteCollection(ctx, deleteValidation, options, listOptions)
+	log := d.Log.WithValues("kind", options.Kind, "resourceVersion", listOptions.ResourceVersion)
+	ctx = klog.NewContext(ctx, log)
+	var method = "delete-collection"
+
+	startLegacy := time.Now()
+	res, err := d.Legacy.DeleteCollection(ctx, deleteValidation, options, listOptions)
+	if err != nil {
+		log.Error(err, "unable to delete collection in legacy storage")
+		d.recordLegacyDuration(true, mode1Str, options.Kind, method, startLegacy)
+		return res, err
+	}
+	d.recordLegacyDuration(false, mode1Str, options.Kind, method, startLegacy)
+
+	go func() {
+		startStorage := time.Now()
+		ctx, cancel := context.WithTimeoutCause(ctx, time.Second*10, errors.New("storage deletecollection timeout"))
+		defer cancel()
+		_, err := d.Storage.DeleteCollection(ctx, deleteValidation, options, listOptions)
+		d.recordStorageDuration(err != nil, mode1Str, options.Kind, method, startStorage)
+	}()
+
+	return res, nil
 }
 
 func (d *DualWriterMode1) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
-	ctx = klog.NewContext(ctx, d.Log)
-	return d.Legacy.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
+	log := d.Log.WithValues("name", name, "kind", options.Kind)
+	ctx = klog.NewContext(ctx, log)
+	var method = "update"
+
+	startLegacy := time.Now()
+	res, async, err := d.Legacy.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
+	if err != nil {
+		log.Error(err, "unable to update in legacy storage")
+		d.recordLegacyDuration(true, mode1Str, options.Kind, method, startLegacy)
+		return res, async, err
+	}
+	d.recordLegacyDuration(false, mode1Str, options.Kind, method, startLegacy)
+
+	updated, err := objInfo.UpdatedObject(ctx, res)
+	if err != nil {
+		log.WithValues("object", updated).Error(err, "could not update or create object")
+	}
+
+	// get the object to be updated
+	foundObj, err := d.Storage.Get(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		log.WithValues("object", foundObj).Error(err, "could not get object to update")
+	}
+
+	// if the object is found, create a new updateWrapper with the object found
+	if foundObj != nil {
+		accessorOld, err := meta.Accessor(foundObj)
+		if err != nil {
+			log.Error(err, "unable to get accessor for original updated object")
+		}
+
+		accessor, err := meta.Accessor(res)
+		if err != nil {
+			log.Error(err, "unable to get accessor for updated object")
+		}
+
+		accessor.SetResourceVersion(accessorOld.GetResourceVersion())
+		accessor.SetUID(accessorOld.GetUID())
+
+		enrichObject(accessorOld, accessor)
+		objInfo = &updateWrapper{
+			upstream: objInfo,
+			updated:  res,
+		}
+	}
```

Comment:
Should we move all of this into the goroutine? Right now we're blocking while we try to read the object from US, and I'm not entirely clear on how we handle the case where we update an object in legacy that isn't in US yet